sdk/clientcore/core/src/main/java/io/clientcore/core/http/models/PagedIterable.java

```diff
@@ -0,0 +1,264 @@
+
+            @Override
+            boolean needToRequestPage() {
+                return (currentPage == null || !currentPage.hasNext()) && nextPage == null;
+            }
+
+            @Override
+            boolean isNextAvailable() {
+                return (currentPage != null && currentPage.hasNext()) || nextPage != null;
+            }
+
+            @Override
+            T getNext() {
+                if ((currentPage == null || !currentPage.hasNext()) && nextPage != null) {
+                    currentPage = nextPage;
+                    nextPage = null;
+                }
+
+                return currentPage.next();
+            }
+
+            @Override
+            void addPage(PagedResponse<T> page) {
+                Iterator<T> pageValues = page.getValue().iterator();
+                if (pageValues.hasNext()) {
+                    nextPage = pageValues;
+                }
+            }
+        };
+    }
+
+    private Iterable<PagedResponse<T>> iterableByPageInternal(PagingOptions pagingOptions) {
+        return () -> new PagedIterator<T, PagedResponse<T>>(pageRetriever, pagingOptions) {
+
+            private PagedResponse<T> nextPage;
+
+            @Override
+            boolean needToRequestPage() {
+                return nextPage == null;
+            }
+
+            @Override
+            boolean isNextAvailable() {
+                return nextPage != null;
+            }
+
+            @Override
+            PagedResponse<T> getNext() {
+                PagedResponse<T> currentPage = nextPage;
+                nextPage = null;
+                return currentPage;
+            }
+
+            @Override
+            void addPage(PagedResponse<T> page) {
+                nextPage = page;
+            }
+        };
+    }
+
+    private abstract static class PagedIterator<T, E> implements Iterator<E> {
+        private static final ClientLogger LOGGER = new ClientLogger(PagedIterator.class);
+
+        private final Function<PagingContext, PagedResponse<T>> pageRetriever;
+        private final Long pageSize;
+        private String continuationToken;
+        private String nextLink;
+        private boolean done;
+
+        PagedIterator(Function<PagingContext, PagedResponse<T>> pageRetriever, PagingOptions pagingOptions) {
+            this.pageRetriever = pageRetriever;
+            this.pageSize = pagingOptions.getPageSize();
+            this.continuationToken = pagingOptions.getContinuationToken();
+        }
+
+        @Override
+        public E next() {
+            if (!hasNext()) {
+                throw LOGGER.logThrowableAsError(new NoSuchElementException("Iterator contains no more elements."));
+            }
+
+            return getNext();
+        }
+
+        @Override
+        public boolean hasNext() {
+            // Request next pages in a loop in case we are returned empty pages for the by item implementation.
+            while (!done && needToRequestPage()) {
```

Comment:
Should this be `if` instead of `while`?

Reply:
The `while` is for a corner case that one page has 0 items, but still have nextLink.

Test case
https://github.com/Azure/azure-sdk-for-java/blob/305218a63cddb23ca045083a4a008a1444be0b08/sdk/clientcore/core/src/test/java/io/clientcore/core/http/models/PagedIterableTests.java#L76-L93